### PR TITLE
Migrating PennTreebank to datapipes

### PR DESCRIPTION
**Pull request by Nayef211**
Reference Issue: #1494

## Summary

- Migrate PennTreebank
- Importing torchdata conditional and raising exception during function call if package is not available

### CODE REVIEW

**Review by erip — CHANGES_REQUESTED** [review]

File: torchtext/datasets/penntreebank.py [hunk 1]
```diff
@@ -1,42 +1,52 @@
...
+
+    url_dp = IterableWrapper([URL[split]])
+    cache_dp = url_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, os.path.basename(x)),
+        hash_dict={os.path.join(root, os.path.basename(URL[split])): MD5[split]},
+        hash_type="md5",
+    )
+    cache_dp = HttpReader(cache_dp).end_caching(mode="w", same_filepath_fn=True)
+    cache_dp = FileLoader(cache_dp, mode="r")
+    return cache_dp.readlines().map(lambda t: t[1][1:-1])
```
[the remark this answers is not in the document]
Comment (Nayef211) [reply torchtext/datasets/penntreebank.py]: Thanks for the feedback! Added a line comment and added the `return_path` kwarg to `readlines`. Lmk if everything else looks okay @parmeet @erip

**Comment by erip** [comment]
LGTM!

File: torchtext/datasets/penntreebank.py [hunk 2]
```diff
@@ -1,42 +1,53 @@
...
+    url_dp = IterableWrapper([URL[split]])
+    cache_dp = url_dp.on_disk_cache(
+        filepath_fn=lambda x: os.path.join(root, os.path.basename(x)),
+        hash_dict={os.path.join(root, os.path.basename(URL[split])): MD5[split]},
+        hash_type="md5",
+    )
+    cache_dp = HttpReader(cache_dp).end_caching(mode="w", same_filepath_fn=True)
+    data_dp = FileOpener(cache_dp, mode="r")
+    # remove single leading and trailing space from the dataset
+    return data_dp.readlines(return_path=False).map(lambda t: t[1:-1])
```
Comment (parmeet) [inline torchtext/datasets/penntreebank.py:53]: nit: can we use `t.strip()` in that case?

Comment (erip) [reply torchtext/datasets/penntreebank.py]: Oooh, I thought `t` was a `List[str]` and 0 and -1 were empty strings. Good catch!

**Pull request merged**